lib/modules/manager/custom/jsonata/utils.ts

```diff
@@ -0,0 +1,99 @@
+import { URL } from 'url';
+import is from '@sindresorhus/is';
+import jsonata from 'jsonata';
+import { logger } from '../../../../logger';
+import * as template from '../../../../util/template';
+import type { PackageDependency } from '../../types';
+import type { JSONataManagerTemplates, JsonataExtractConfig } from './types';
+
+export const validMatchFields = [
+  'depName',
+  'packageName',
+  'currentValue',
+  'currentDigest',
+  'datasource',
+  'versioning',
+  'extractVersion',
+  'registryUrl',
+  'depType',
+] as const;
+
+type ValidMatchFields = (typeof validMatchFields)[number];
+
+export async function handleMatching(
+  json: unknown,
+  packageFile: string,
+  config: JsonataExtractConfig,
+): Promise<PackageDependency[]> {
+  // Pre-compile all JSONata expressions once
+  const compiledExpressions = config.matchStrings
+    .map((query) => {
+      try {
+        return jsonata(query);
+      } catch (err) {
+        logger.warn({ query, err }, 'Failed to compile JSONata query');
+        return null;
+      }
+    })
+    .filter((expr) => expr !== null);
+
+  // Execute all expressions in parallel
+  const results = await Promise.all(
```

Comment:
This is implicitly any, parse this object with a Zod schema like we [do for the `custom` datasource](https://github.com/renovatebot/renovate/blob/2f805f4b7d746ffbe0c6509be592ae0df24979f4/lib/modules/datasource/custom/index.ts#L71)